### DISCUSSION TITLE
How to apply continuity boundary condition in a composite domain/geometry

### DISCUSSION THREAD

**Discussion by ma-sadeghi** [lululxvi/deepxde#848] [Q&A]
# Problem statement

Consider the steady state heat equation on a 2D rectangular domain (unit square) that is composed of two different materials with heat conductivity coefficients $k_1$ and $k_2$. The governing PDE is:

$$
k(x) \Big[ \frac{\partial^2 u}{\partial x^2} + \frac{\partial^2 u}{\partial y^2} \Big] = 0
$$

where $k(x)$ is defined as:

$$
\begin{align}
0 < x < 0.5 \quad ; \quad k = k_1 \\ 
0.5 < x < 1 \quad ; \quad k = k_2
\end{align}
$$

with boundary conditions defined as:

$$
\begin{align}
x = 0 \quad ; \quad u = 1 \\
x = 1 \quad ; \quad u = 0 \\
y = 0 \quad ; \quad \frac{\partial u}{\partial y} = 0 \\
y = 1 \quad ; \quad \frac{\partial u}{\partial y} = 0
\end{align}
$$

Finally, for the coupling between the two domains, we have the following _internal_ boundary condition:

$$
x = 0.5 \quad ; \quad k_1 \frac{\partial u}{\partial x} = k_2 \frac{\partial u}{\partial x}
$$

# My question
Now, my question is: how can I enforce the above condition (continuity at the internal boundary)? I'd really appreciate it if you could help me with this!

# Additional information
Btw, here's the code for the above problem, but it doesn't capture the presence of two different materials, and I think that's expected, since the coupling between the two domains (i.e., the last boundary condition) is missing).

Another point to note is that although I've included a variable conductivity coefficient in the `pde`, the PINN solver still can't capture the presence of two different materials, but again that's I think expected, because the PDE is at steady state, so the `k` is just a multiplier and so the optimizer ignores it altogether.

```python
import tensorflow as tf
import deepxde as dde

k1, k2 = 1.0, 10.0
xmin, ymin = 0.0, 0.0
xmax, ymax = 1.0, 1.0

def pde(x, u):
    dy_xx = dde.grad.hessian(u, x, i=0, j=0)
    dy_yy = dde.grad.hessian(u, x, i=1, j=1)
    k = tf.where(x[:, 0] < 0.5, k1, k2)
    return k * (dy_xx + dy_yy)

geom = dde.geometry.Rectangle([xmin, ymin], [xmax, ymax])

def boundary_top(x, on_boundary):
    return on_boundary and np.isclose(x[1], 1)

def boundary_bottom(x, on_boundary):
    return on_boundary and np.isclose(x[1], 0)

def boundary_right(x, on_boundary):
    return on_boundary and np.isclose(x[0], 1)

def boundary_left(x, on_boundary):
    return on_boundary and np.isclose(x[0], 0)

bc_left = dde.DirichletBC(geom, lambda x: np.ones((len(x), 1))*1.0, boundary_left)
bc_right = dde.DirichletBC(geom, lambda x: np.ones((len(x), 1))*0.0, boundary_right)
bc_top = dde.NeumannBC(geom, lambda x: np.zeros((len(x), 1)), boundary_top)
bc_bottom = dde.NeumannBC(geom, lambda x: np.zeros((len(x), 1)), boundary_bottom)

data = dde.data.PDE(geom,
                    pde,
                    [bc_left, bc_right, bc_top, bc_bottom],
                    num_domain=1000,
                    num_boundary=100)

net = dde.nn.FNN([2] + [50] * 4 + [1], "tanh", "Glorot uniform")
model = dde.Model(data, net)

model.compile("adam", lr=0.001)
losshistory, train_state = model.train(iterations=10_000)
dde.saveplot(losshistory, train_state, issave=False, isplot=True)
```

**Comment by forxltk** [comment]
1. Sample enough points in `x=0.5`, and take them as anchors in `dde.data.PDE`.
2. Define where the internal boundary is.  
```
def boundary_internal(x, on_boundary):
    return np.isclose(x[0], 0.5)
 ```
3. Use `OperatorBC` to define the condition.

**Comment by juliosdutra** [comment]
Hi, Amin Sadeghi.

Have you resolved the issue? Could you please share the final code with me?

Thank you.

**Reply by ma-sadeghi** [reply]
Hi @juliosdutra. I got distracted with other projects, so didn't get to work on this one. Would be happy to hear from your progress if you're still working on it!

**Reply by juliosdutra** [reply]
Hi, @ma-sadeghi. You can check on:  https://github.com/lululxvi/deepxde/issues/447#issuecomment-1443807965 Any suggestions/comments are welcome.
